File: openpearl-code/compiler/src/org/openpearl/compiler/DeadLockDetection/GraphUtils/Graph.java
/*
 * [A "BSD license"]
 *  Copyright (c) 2022 Jan Knoblauch
 *  
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *
 *  1. Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *  2. Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *  3. The name of the author may not be used to endorse or promote products
 *     derived from this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 *  IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 *  OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 *  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 *  NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 *  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 *  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 *  THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */


package org.openpearl.compiler.DeadLockDetection.GraphUtils;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Graph<T extends IGraphNode>
{
    private final IntegerGraph integerGraph = new IntegerGraph();

    private final Map<String, T> nodesById = new HashMap<>();
    private final List<String> idIndexes = new ArrayList<>();
    private int autoIncrementIndex = 0;

    private int addWithIndex(T node)
    {
        int result = -1;

        if (!nodesById.containsKey(node.getId()))
        {
            result = autoIncrementIndex;
            autoIncrementIndex++;
            nodesById.put(node.getId(), node);
            idIndexes.add(node.getId());
        }
        else
        {
            result = idIndexes.indexOf(node.getId());
        }

        return result;
    }

    private String getIdentifier(T node)
    {
        addWithIndex(node);

        return node.getId();
    }

    private int getIndex(T node)
    {
        return getIndex(getIdentifier(node));
    }

    private int getIndex(String nodeId)
    {
        if (nodeId == null)
        {
            return -1;
        }

        return idIndexes.indexOf(nodeId);
    }

    public T get(String nodeIdentifier)
    {
        if ((nodeIdentifier == null) || (!nodesById.containsKey(nodeIdentifier)))
        {
            return null;
        }

        return nodesById.get(nodeIdentifier);
    }

    public T get(int nodeIndex)
    {
        return get(idIndexes.get(nodeIndex));
    }

    public void addEdge(Edge edge)
    {
        addEdge((T)edge.source, (T)edge.target, edge.label);
    }

    public void addEdge(T sourceNode, T targetNode, String label)
    {
        int sourceIndex = addWithIndex(sourceNode);
        int targetIndex = addWithIndex(targetNode);

        integerGraph.addEdge(sourceIndex, targetIndex, label);
    }

    public void removeEdge(T sourceNode, T targetNode)
    {
        int sourceIndex = getIndex(sourceNode);
        int targetIndex = getIndex(targetNode);

        integerGraph.removeEdge(sourceIndex, targetIndex);
    }

    public void removeNode(T node)
    {
        int nodeIndex = getIndex(node);

        integerGraph.removeNode(nodeIndex);
    }

    public List<T> getNodesTopologicalSorted()
    {
        List<Integer> indexes = integerGraph.topologicalSort();

        List<T> nodes = new ArrayList<>();

        for (Integer nodeIndex : indexes)
        {
            nodes.add(get(nodeIndex));
        }

        return nodes;
    }

    public void printGraph()
    {
        for (Map.Entry<Integer, List<Integer>> nodeEntry : integerGraph.getEdges().entrySet())
        {
            String sourceNodeIndex = idIndexes.get(nodeEntry.getKey());

            T sourceNode = nodesById.get(sourceNodeIndex);

            System.out.print(sourceNode.getId() + " -> ");

            for (Integer targetNodeIndex : nodeEntry.getValue())
            {
                String targetNodeIdentifier = idIndexes.get(targetNodeIndex);

                T targetNode = nodesById.get(targetNodeIdentifier);

                System.out.print(targetNode.getId() + " " + targetNode.getLabel() + ", ");
            }

            System.out.println("");
        }
    }

    public List<List<T>> getPathsByLabel(T startNode, String label)
    {
        List<List<T>> resultPaths = new ArrayList<>();

        if (startNode == null)
        {
            return resultPaths;
        }

        if (!containsNode(startNode))
        {
            return resultPaths;
        }

        for (List<Integer> currentPath : integerGraph.getPathsByLabel(getIndex(startNode), label))
        {
            resultPaths.add(getNodesByIndex(currentPath));
        }

        return resultPaths;
    }

    public List<T> getCycle()
    {
        return getNodesByIndex(integerGraph.getCycle());
    }

    public List<T> getStartNodes()
    {
        return getNodesByIndex(integerGraph.getStartNodes());
    }

    public List<T> getExclusivePathFromNode(T node, boolean allowMultipleSubPaths)
    {
        return getNodesByIndex(integerGraph.getExclusivePathFromNode(getIndex(node), allowMultipleSubPaths));
    }

    private List<T> getNodesByIndex(List<Integer> nodeIndexes)
    {
        List<T> nodes = new ArrayList<>();

        if (nodeIndexes.size() == 0)
        {
            return nodes;
        }

        for (int index : nodeIndexes)
        {
            nodes.add(get(index));
        }

        return nodes;
    }

    public boolean containsEdge(T source, T target, String label)
    {
        int sourceIndex = getIndex(source);
        int targetIndex = getIndex(target);

        return integerGraph.isConnected(sourceIndex, targetIndex, label);
    }

    public boolean containsNode(T node)
    {
        if (node == null)
        {
            return false;
        }

        return nodesById.containsKey(node.getId());
    }
}

File: openpearl-code/runtime/common/DynamicDeadlockDetection/Graph.cc
/*
 * [A "BSD license"]
 *  Copyright (c) 2022 Jan Knoblauch
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *
 *  1. Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *  2. Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *  3. The name of the author may not be used to endorse or promote products
 *     derived from this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 *  IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 *  OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 *  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 *  NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 *  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 *  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 *  THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */


#include "../DynamicDeadlockDetection/Graph.h"

namespace pearlrt {

   void Graph::performDFS(int currentNode, int visitedNodes[])
   {
      cycleDetectionPathStack.push(currentNode);

      if (visitedNodes[currentNode] == NODE_STATE_IS_BEING_VISITED)
      {
         vector<int> cycleDetectionPathVector;

         while (!cycleDetectionPathStack.empty())
         {
            cycleDetectionPathVector.push_back(cycleDetectionPathStack.top());
            cycleDetectionPathStack.pop();
         }

         reverse(cycleDetectionPathVector.begin(), cycleDetectionPathVector.end());

         int lastCycleNode = cycleDetectionPathVector[cycleDetectionPathVector.size() - 1];
         bool cycleStarted = false;

         for (int currentCycleNode : cycleDetectionPathVector)
         {
            if (!cycleStarted)
            {
               if (currentCycleNode == lastCycleNode)
               {
                  cycleStarted = true;
               }
            }

            if (cycleStarted)
            {
               cycleDetectionPath.push_back(getNodeIdentifier(currentCycleNode));
            }
         }

         cycleDetected = true;

         return;
      }

      visitedNodes[currentNode] = NODE_STATE_IS_BEING_VISITED;

      for (int reachableNode : graph[currentNode])
      {
         if (visitedNodes[reachableNode] != NODE_STATE_FINISHED_VISITING)
         {
            performDFS(reachableNode, visitedNodes);
         }
      }

      visitedNodes[currentNode] = NODE_STATE_FINISHED_VISITING;

      if (!cycleDetectionPathStack.empty())
      {
         cycleDetectionPathStack.pop();
      }
   }

   int Graph::addNode(const string& nodeIdentifier)
   {
      int nodeId = getNodeId(nodeIdentifier);

      if (nodeId >= 0)
      {
         return nodeId;
      }

      nodeIdentifiers.push_back(nodeIdentifier);

      nodeId = nodeCount;

      nodeCount++;

      return nodeId;
   }

   int Graph::getNodeId(const string& nodeIdentifier)
   {
      int index = 0;

      for (const string& currentIdentifier : nodeIdentifiers)
      {
         if (nodeIdentifier == currentIdentifier)
         {
            return index;
         }

         index++;
      }

      return -1;
   }

   string Graph::getNodeIdentifier(int nodeId)
   {
      if ((nodeId < 0) || (nodeId >= nodeCount))
      {
         return "";
      }

      return nodeIdentifiers[nodeId];
   }

   void Graph::addEdge(const string& sourceNodeIdentifier, const string& targetNodeIdentifier)
   {
      int sourceId = addNode(sourceNodeIdentifier);
      int targetId = addNode(targetNodeIdentifier);

      if (find(graph[sourceId].begin(), graph[sourceId].end(), targetId) == graph[sourceId].end())
      {
         graph[sourceId].push_back(targetId);
      }
   }

   void Graph::removeEdge(const string& sourceNodeIdentifier, const string& targetNodeIdentifier)
   {
      int sourceId = getNodeId(sourceNodeIdentifier);
      int targetId = getNodeId(targetNodeIdentifier);

      if ((sourceId < 0) || (targetId < 0))
      {
         return;
      }

      graph[sourceId].erase(remove(graph[sourceId].begin(), graph[sourceId].end(), targetId), graph[sourceId].end());
   }

   vector<string> Graph::getCycle()
   {
      while (!cycleDetectionPathStack.empty())
      {
         cycleDetectionPathStack.pop();
      }

      cycleDetectionPath.clear();

      cycleDetected = false;

      int visited[graphNodeCount] = {};

      for (int i = 0; i < nodeCount; i++)
      {
         if (visited[i] == NODE_STATE_NOT_VISITED)
         {
            performDFS(i, visited);
         }

         if (cycleDetected)
         {
            break;
         }
      }

      return cycleDetectionPath;
   }

   void Graph::printGraph()
   {
      for (int nodeId = 0; nodeId < nodeCount; nodeId++)
      {
         cout << "Node " << getNodeIdentifier(nodeId) << ": ";

         for (auto nodeEdge : graph[nodeId])
         {
            cout << getNodeIdentifier(nodeEdge) << ", ";
         }

         cout << endl;
      }
   }

   string Graph::getFirstEdge(const string& nodeIdentifier)
   {
      int nodeId = getNodeId(nodeIdentifier);

      if (nodeId < 0)
      {
         return "";
      }

      for (auto nodeEdge : graph[nodeId])
      {
         return getNodeIdentifier(nodeEdge);
      }

      return "";
   }

   vector<string> Graph::getOutgoingEdges(const string& nodeIdentifier)
   {
      vector<string> outgoingNodes;

      int nodeId = getNodeId(nodeIdentifier);

      if (nodeId < 0)
      {
         return outgoingNodes;
      }

      for (auto nodeEdge : graph[nodeId])
      {
         outgoingNodes.push_back(getNodeIdentifier(nodeEdge));
      }

      return outgoingNodes;
   }

   vector<string> Graph::getIncomingEdges(const string& nodeIdentifier)
   {
      vector<string> incomingNodes;

      int targetNodeId = getNodeId(nodeIdentifier);

      if (targetNodeId < 0)
      {
         return incomingNodes;
      }

      for (int currentNodeId = 0; currentNodeId < nodeCount; currentNodeId++)
      {
         for (auto nodeEdge : graph[currentNodeId])
         {
            if (nodeEdge == targetNodeId)
            {
               incomingNodes.push_back(getNodeIdentifier(currentNodeId));

               goto outerLoop;
            }
         }

         outerLoop:;
      }

      return incomingNodes;
   }
}
